### PR TITLE
applications: asset_tracker: Convert ot new k_work API

### PR DESCRIPTION
**Pull request by jhn-nordic**
Convert to new k_work API due to deprecation.

Jira: NCSDK-9409

Signed-off-by: Jon Helge Nistad <jon.helge.nistad@nordicsemi.no>

link to best practices: https://github.com/zephyrproject-rtos/zephyr/issues/33104

### CODE REVIEW

File: applications/asset_tracker/src/env_sensors/env_sensors.c
Comment (jhn-nordic) [inline applications/asset_tracker/src/env_sensors/env_sensors.c:218]: @rlubos In these situations. Should we just throw away the "reamaning_get" and cancel anyway?

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [inline applications/asset_tracker/src/main.c:570]: There are lots of these related to _cloud_reboot_work_ . Are there other consequences than an extra reboot if the cancel fails?

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [inline applications/asset_tracker/src/main.c:991]: This I believe need to be synchronized by other mechanisms

File: applications/asset_tracker_v2/src/led/led_pwm.c
Comment (jhn-nordic) [inline applications/asset_tracker_v2/src/led/led_pwm.c:123]: we need to look more into this one

File: applications/asset_tracker_v2/src/led/led_pwm.c
Comment (jhn-nordic) [inline applications/asset_tracker_v2/src/led/led_pwm.c:175]: This is needs investigation

File: applications/asset_tracker_v2/src/modules/cloud_module.c
Comment (jhn-nordic) [inline applications/asset_tracker_v2/src/modules/cloud_module.c:476]: what happpens if not cancelled?

File: applications/asset_tracker_v2/src/modules/cloud_module.c
Comment (jhn-nordic) [inline applications/asset_tracker_v2/src/modules/cloud_module.c:558]: what happpens if not cancelled?

File: applications/asset_tracker_v2/src/modules/data_module.c
Comment (jhn-nordic) [inline applications/asset_tracker_v2/src/modules/data_module.c:665]: Is this one safe?

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/main.c:570]: @jayteemo I think you have most knowledge about the reboot mechanisms in the asset tracker.

File: applications/asset_tracker/src/env_sensors/env_sensors.c
Comment (trond-snekvik) [reply applications/asset_tracker/src/env_sensors/env_sensors.c:218]: FWIW, I think you should. A safer version of this check already exists inside the cancel function.

File: applications/asset_tracker/src/env_sensors/bsec.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/env_sensors/bsec.c:372]: What happens if this fails?

File: applications/asset_tracker/src/gps_controller/gps_controller.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/gps_controller/gps_controller.c:105]: What happens if this fails? Would this be simpler if we just had one work item and used a flag to determine whether to start or stop?

File: applications/asset_tracker/src/main.c
Comment (trond-snekvik) [reply applications/asset_tracker/src/main.c:991]: Should use `k_work_schedule_for_queue` if we don't want to restart a running timer

File: applications/asset_tracker/src/main.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/main.c:1512]: If this fails, we'll unconditionally shut down. Perhaps the reboot work handler could check the state variable before shutting down?

File: applications/asset_tracker/src/main.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/main.c:1804]: What happens if this fails?

File: applications/asset_tracker/src/ui/led_pwm.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/ui/led_pwm.c:139]: Can be `schedule`

File: applications/asset_tracker/src/ui/led_pwm.c
Comment (trond-snekvik) [inline applications/asset_tracker/src/ui/led_pwm.c:145]: If this fails, and step count isn't > 0, what happens in the work_handler?

File: applications/asset_tracker_v2/src/led/led.c
Comment (trond-snekvik) [inline applications/asset_tracker_v2/src/led/led.c:106]: Can be `schedule()`, same goes for calls in the handler

File: applications/asset_tracker/src/env_sensors/env_sensors.c
Comment (rlubos) [reply applications/asset_tracker/src/env_sensors/env_sensors.c:218]: I agree with @trond-snekvik

File: applications/asset_tracker_v2/src/watchdog/watchdog.c
Comment (trond-snekvik) [inline applications/asset_tracker_v2/src/watchdog/watchdog.c:95]: Can be a `schedule`

File: applications/asset_tracker/src/main.c
Comment (rlubos) [reply applications/asset_tracker/src/main.c:570]: Well in general the cancelation can fail if the canceled work is already being processed by the work thread. I don't see a way to prevent the work from being executed at this point (other than aborting the thread, which is not what we want I guess).

Comment (rlubos) [reply applications/asset_tracker/src/main.c:570]: Btw, if you're concerned about syncrhnization issues, when the cancelled work is already being exectured, there's `k_work_cancel_delayable_sync()` which should block until the work is finished (in case it's already being handled by the work thread).

Comment (jayteemo) [reply applications/asset_tracker/src/main.c:570]: > @jayteemo I think you have most knowledge about the reboot mechanisms in the asset tracker.

The reboot work is in asset tracker is just used as a fail safe for connect/disconnect problems.

File: applications/asset_tracker/src/main.c
Comment (jayteemo) [reply applications/asset_tracker/src/main.c:1804]: the gps enabled state will be toggled if the work fails to be canceled

File: applications/asset_tracker/src/env_sensors/bsec.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/env_sensors/bsec.c:372]: the work item will return early if data_send_interval_s==0

File: applications/asset_tracker/src/env_sensors/env_sensors.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/env_sensors/env_sensors.c:218]: done

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/main.c:1512]: There is no good state variable to use in this case. We can live with a reboot in corner cases. More risky to add state handling to this application

File: applications/asset_tracker_v2/src/modules/data_module.c
Comment (jhn-nordic) [reply applications/asset_tracker_v2/src/modules/data_module.c:665]: We can live with a failing cancel here

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/main.c:570]: keeping as is

File: applications/asset_tracker/src/env_sensors/bsec.c
Comment (pabigot) [reply applications/asset_tracker/src/env_sensors/bsec.c:372]: please add a comment recording that so future maintainers don't need to analyze the code to find out.

File: applications/asset_tracker/src/env_sensors/env_sensors.c
Comment (pabigot) [inline applications/asset_tracker/src/env_sensors/env_sensors.c:218]: Please document in a comment what happens if this does not complete immediately so the work continues to run.

File: applications/asset_tracker/src/main.c
Comment (pabigot) [inline applications/asset_tracker/src/main.c:991]: What's the value in checking the remaining time?  Just schedule it; if it's already scheduled (remaining is not zero) there will be no change to the existing deadline.

**Review by pabigot — COMMENTED** [review]
There are enough unchecked cancels and mixes of schedule vs reschedule I can't confirm this is correct without a deep analysis.

I have only one substantial comment related to a use of schedule conditional on a check of remaining time.   Such checks are racy and should be avoided.  https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#don-t-optimize-prematurely

File: applications/asset_tracker/src/main.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/main.c]: done

File: applications/asset_tracker_v2/src/watchdog/watchdog.c
Comment (jhn-nordic) [reply applications/asset_tracker_v2/src/watchdog/watchdog.c:95]: done

File: applications/asset_tracker_v2/src/led/led.c
Comment (jhn-nordic) [reply applications/asset_tracker_v2/src/led/led.c]: done

File: applications/asset_tracker/src/env_sensors/bsec.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/env_sensors/bsec.c:373]: done

File: applications/asset_tracker/src/gps_controller/gps_controller.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/gps_controller/gps_controller.c:106]: considered it to be safe since they run on same queue

File: applications/asset_tracker_v2/src/watchdog/watchdog.c
Comment (trond-snekvik) [inline applications/asset_tracker_v2/src/watchdog/watchdog.c:96]: Same as in the button handler, I believe this is no longer possible in normal scenarios

**Review by trond-snekvik — APPROVED** [review]
One minor comment. It'd be helpful if the remaining cancels were documented as well, including the ones we're accepting, but they already have comments from others, so I'll approve

File: applications/asset_tracker_v2/src/watchdog/watchdog.c
Comment (jhn-nordic) [reply applications/asset_tracker_v2/src/watchdog/watchdog.c:96]: fixed

File: applications/asset_tracker/src/ui/led_pwm.c
Comment (pabigot) [inline applications/asset_tracker/src/ui/led_pwm.c:146]: This isn't new, so shouldn't block the change, but this doesn't look right.  The check at 151 suggests that disabling led updates involves setting `led->effect` to NULL.  However, the work handler unconditionally dereferences `led->effect`.  If the cancel fails (or if whatever sets `led->effect` to NULL has another race with the work handler) the handler could cause a fault.

**Review by pabigot — COMMENTED** [review]
I think there's an existing bug/race condition in led management (assuming I'm right in inferring how led updates are disabled) but I don't see anything in the change that introduces new problems.  I can approve if my inference is wrong or the problem will be addressed elsewhere.

File: applications/asset_tracker/src/ui/led_pwm.c
Comment (jhn-nordic) [reply applications/asset_tracker/src/ui/led_pwm.c:146]: Thanks. I have added  a new task to investigated this: CIA-317

**Review by pabigot — APPROVED** [review]

**Review by jtguggedal — APPROVED** [review]

**Review by balaji-nordic — COMMENTED** [review]
Baring alignment issues, this PR looks okay.

**Pull request merged**
